monai/data/dataset.py

```diff
@@ -50,13 +55,120 @@ def __getitem__(self, index):
+        return item_transformed
+
+    def _pre_first_random_transformed(self, item_transformed):
+        """
+            A function to cache the expensive input data transform operations
+            so that huge data sets (larger than computer memory) can be processed
+            on the fly as needed, and intermediate results written to disk for
+            future use.
+        Args:
+            item_transformed: The current data element to be mutated into transformed representation
+
+        Returns:
+            The transformed data_element, either from cache, or explicitly computing it.
+
+        Warning:
+            The current implementation does not encode transform information as part of the
+            hashing mechanism used for generating cache names.  If the transforms applied are
+            changed in any way, the objects in the cache dir will be invalid.  The hash for the
+            cache is ONLY dependant on the input filename paths.
+        """
+        if item_transformed.get("cached", False) is False:
+            hashfile = None
+            if self.cache_dir is not None:
+                cache_dir_path: Path = Path(self.cache_dir)
+                if cache_dir_path.is_dir():
+                    # TODO: Find way to hash transforms content as part of the cache
+                    data_item_md5 = hashlib.md5(json.dumps(item_transformed, sort_keys=True).encode('utf-8')).hexdigest()
+                    hashfile: Path = Path(cache_dir_path) / f"{data_item_md5}.pt"
+
+            if hashfile is not None and hashfile.is_file():
+                item_transformed = torch.load(hashfile)
+            else:
+                item_transformed = self._pre_first_random_transform(item_transformed)
+                if hashfile is not None:
+                    # add sentinal flag to indicate that the transforms have already been computed.
+                    item_transformed["cache"] = True
+                    # NOTE: Writing to ".temp_write_cache" and then using a nearly atomic rename operation
+                    #       to make the cache more robust to manual killing of parent process
+                    #       which may leave partially written cache files in an incomplete state
+                    temp_hash_file: Path = hashfile.with_suffix(".temp_write_cache")
+                    torch.save(item_transformed, temp_hash_file)
+                    temp_hash_file.rename(hashfile)
+
+        return item_transformed
+
+    def __getitem__(self, index):
+        pre_random_item = self._pre_first_random_transformed(self.data[index])
+        return self._first_random_and_beyond_transform(pre_random_item)
+
+
 class CacheDataset(Dataset):
     """
-    Dataset with cache mechanism that can load data and cache deterministic transforms' result during training.
+    Dataset with an in-memory cache mechanism that can load data and cache deterministic transforms' result during
```

Comment:
The CacheDataset should be restored to the original implementation as it doesn't inherit PersistentDataset.
Thanks.